### ISSUE TITLE
Encoder becomes corrupted when it encounters an error

### ISSUE THREAD

**Issue by myronmarston** [beatrichartz/csv#12]
Playing around with your library in iex yielded some pretty odd results:

``` iex
iex(1)> ["a"] |> CSV.encode |> CSV.decode |> Enum.to_list
** (Protocol.UndefinedError) protocol Enumerable not implemented for "a"
    (elixir) lib/enum.ex:1: Enumerable.impl_for!/1
    (elixir) lib/enum.ex:112: Enumerable.reduce/3
    (elixir) lib/enum.ex:981: Enum.map/2
             lib/csv/encoder.ex:47: CSV.Encoder.encode_row/3
             lib/csv/encoder.ex:42: anonymous fn/4 in CSV.Encoder.encode/2
    (elixir) lib/stream.ex:650: Stream.do_transform/7
    (elixir) lib/enum.ex:1740: Enum.take/2
             lib/csv/decoder.ex:120: CSV.Decoder.produce/2
iex(1)> [["b"]] |> CSV.encode |> CSV.decode |> Enum.to_list
[]
iex(2)> [["b"]] |> CSV.encode |> CSV.decode |> Enum.to_list
[["b"], ["b"]]
iex(3)> [["b"]] |> CSV.encode |> CSV.decode |> Enum.to_list
[]
iex(4)> [["b"]] |> CSV.encode |> CSV.decode |> Enum.to_list
[["b"]]
```

Notice that `[["b"]] |> CSV.encode |> CSV.decode |> Enum.to_list` resulted in `[]` the first time, then `[["b"], ["b"]]` and `[]`, and then finally, on the 4th try, `[["b"]]`, as expected.  The mistake made initially causes the encoder to become corrupted apparently.  If exit the iex session and re-do it without the bad line the problem does not occur.

**Comment by beatrichartz** [comment]
Thanks for reporting, this seems like an interesting one, will have a look into it later today.

**Comment by beatrichartz** [comment]
The stream producer now gracefully recovers from errors - did you find any other cases of this happening? If not I'll close this. Fix is released in `1.1.1`.

**Comment by myronmarston** [comment]
Thanks.  I'll let you know if I run into any more cases like this.
